double bump probing as a feature

AnHardt opened the pull request:
Why double touch probing is not a good thing.

It's widely believed we can get better **probing** results when using a double touch when probing.

Let's compare to double touch **homing**.
Or better let's begin with single touch **homing**.
We home to find out our position, so our position is unknown.
To find the endstop we have to move into the direction of the endstop.
The maximum way we have to move is a bit longer than the axis length.
When we arrive at the endstop - when it triggers, the stepper pulses are stopped immediately.
It's a sudden stop. No smooth deacceleration is possible.
Depending on the speed and the moving mass we lose steps here.
Only if we approached slow enough (below jerk speed?) we will not lose steps.

Moving a complete axis length, that slow, takes for ever.
To speed up homing, we now make the first approach faster, get a guess about our position,
back up a bit and make a second slower approach to get a exact result without losing steps.

What we do in double touch probing is the same. But the difference here is:
a. we already know where we are
b. if the first approach is too fast we will lose steps here too.
But this time there is no second approach to set the position to 0. We are measuring only.
The lost steps are permanent until we home the next time.

So if you experienced permanently rising values in M48, you now know why. (Too fast, suddenly stopped, first approach)

What can we do to improve probing?
We can use the information about our current position.
We can make a really fast, but decelerated, move **to a place we know** it is a bit before the trigger point. And then move the rest of the way really slow.

The result is: 
We can make a faster first approach.
The way is a little shorter.
No steps can got lost during stopping the first approach.

This PR makes `PROBE_DOUBLE_TOUCH` optional and switched off by default.

If you are convinced by my theory, we can drop the double touch code at all and save the new configuration option.

brainscan commented:
This makes a lot of sense to me.

thinkyhead commented:
@brainscan Have you got a copy of [CodeHub](https://github.com/thedillonb/CodeHub) on your iPhone? I use it on my iPad and I find it's a great way to keep up with projects.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -2110,16 +2110,21 @@ static void clean_up_after_endstop_or_probe_move() {
+      // move up the retract distance
+      do_blocking_move_to_z(current_position[Z_AXIS] + home_bump_mm(Z_AXIS), Z_PROBE_SPEED_FAST);
+    #else
+      // move down fast, cloth to the bed
```
"close" (from Old French _clos_) as opposed to "cloth" (O.E. _clāth_, related to the German _Kleid_).

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -2110,16 +2110,21 @@ static void clean_up_after_endstop_or_probe_move() {
+      do_blocking_move_to_z(current_position[Z_AXIS] + home_bump_mm(Z_AXIS), Z_PROBE_SPEED_FAST);
+    #else
+      // move down fast, cloth to the bed
+      do_blocking_move_to_z(home_bump_mm(Z_AXIS), Z_PROBE_SPEED_FAST);
```
This actually appears to be a raise.

AnHardt replied:
It's neither a lowering nor a rising is a go to a absolute position.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -2110,16 +2110,21 @@ static void clean_up_after_endstop_or_probe_move() {
-    // move back down slowly to find bed
-    do_blocking_move_to_z(current_position[Z_AXIS] - home_bump_mm(Z_AXIS) * 2, Z_PROBE_SPEED_SLOW);
+    // move down slowly to find bed
+    do_blocking_move_to_z(current_position[Z_AXIS] -2.0*home_bump_mm(Z_AXIS), Z_PROBE_SPEED_SLOW);
```
Ah good! The "single probe" option uses the slow speed. 😄

thinkyhead commented on Marlin/example_configurations/Cartesio/Configuration.h:
```diff
@@ -491,6 +491,8 @@
 #define Z_PROBE_SPEED_FAST HOMING_FEEDRATE_Z
 // Speed for the second approach when probing
```
I guess these comments need to be qualified now…

``` cpp
// Speed for the first approach when double-probing (with PROBE_DOUBLE_TOUCH)
#define Z_PROBE_SPEED_FAST HOMING_FEEDRATE_Z
```

``` cpp
// Speed for the "accurate" probe of each point
#define Z_PROBE_SPEED_SLOW (Z_PROBE_SPEED_FAST / 2)
```

thinkyhead commented:
I support this 💯  percent.

jbrazio commented:
:shipit:

AnHardt committed:
double bump probing as a feature

Why double touch probing is not a good thing.

It's widely believed we can get better __probing__ results when using a double touch when probing.

Let's compare to double touch __homing__.
Or better let's begin with single touch __homing__.
We home to find out out position, so our position is unknown.
To find the endstop we have to move into the direction of the endstop.
The maximum way we have to move is a bit longer than the axis length.
When we arrive at the endstop - when it triggers, the stepper pulses are stopped immediately.
It's a sudden stop. No smooth deacceleration is possible.
Depending on the speed and the moving mass we lose steps here.
Only if we approached slow enough (below jerk speed?) we will not lose steps.

Moving a complete axis length, that slow, takes for ever.
To speed up homing, we now make the first approach faster, get a guess about our position,
back up a bit and make a second slower approach to get a exact result without losing steps.

What we do in double touch probing is the same. But the difference here is:
a. we already know where we are
b. if the first approach is to fast we will lose steps here to.
But this time there is no second approach to set the position to 0. We are measuring only.
The lost steps are permanent until we home the next time.

So if you experienced permanently rising values in M48 you now know why. (Too fast, suddenly stopped, first approach)

What can we do to improve probing?
We can use the information about our current position.
We can make a really fast, but deaccelerated, move to a place we know it is a bit before the trigger point.
And then move the rest of the way really slow.

brainscan commented:
@Thinkyhead I do now, thanks 😀
